MoltenVK/MoltenVK/GPUObjects/MVKPipeline.mm

```diff
@@ -557,26 +539,81 @@
+		case VK_DYNAMIC_STATE_DEPTH_BIAS:                  return DepthBias;
+		case VK_DYNAMIC_STATE_DEPTH_BIAS_ENABLE:           return DepthBiasEnable;
+		case VK_DYNAMIC_STATE_DEPTH_BOUNDS:                return DepthBounds;
+		case VK_DYNAMIC_STATE_DEPTH_BOUNDS_TEST_ENABLE:    return DepthBoundsTestEnable;
+		case VK_DYNAMIC_STATE_DEPTH_CLAMP_ENABLE_EXT:      return DepthClipEnable;
+		case VK_DYNAMIC_STATE_DEPTH_CLIP_ENABLE_EXT:       return DepthClipEnable;
+		case VK_DYNAMIC_STATE_DEPTH_COMPARE_OP:            return DepthCompareOp;
+		case VK_DYNAMIC_STATE_DEPTH_TEST_ENABLE:           return DepthTestEnable;
+		case VK_DYNAMIC_STATE_DEPTH_WRITE_ENABLE:          return DepthWriteEnable;
+		case VK_DYNAMIC_STATE_FRONT_FACE:                  return FrontFace;
+		case VK_DYNAMIC_STATE_LOGIC_OP_EXT:                return LogicOp;
+		case VK_DYNAMIC_STATE_LOGIC_OP_ENABLE_EXT:         return LogicOpEnable;
+		case VK_DYNAMIC_STATE_PATCH_CONTROL_POINTS_EXT:    return PatchControlPoints;
+		case VK_DYNAMIC_STATE_POLYGON_MODE_EXT:            return PolygonMode;
+		case VK_DYNAMIC_STATE_PRIMITIVE_RESTART_ENABLE:    return PrimitiveRestartEnable;
+		case VK_DYNAMIC_STATE_PRIMITIVE_TOPOLOGY:          return PrimitiveTopology;
+		case VK_DYNAMIC_STATE_RASTERIZER_DISCARD_ENABLE:   return RasterizerDiscardEnable;
+		case VK_DYNAMIC_STATE_SCISSOR:                     return Scissors;
+		case VK_DYNAMIC_STATE_SCISSOR_WITH_COUNT:          return Scissors;
+		case VK_DYNAMIC_STATE_STENCIL_COMPARE_MASK:        return StencilCompareMask;
+		case VK_DYNAMIC_STATE_STENCIL_OP:                  return StencilOp;
+		case VK_DYNAMIC_STATE_STENCIL_REFERENCE:           return StencilReference;
+		case VK_DYNAMIC_STATE_STENCIL_TEST_ENABLE:         return StencilTestEnable;
+		case VK_DYNAMIC_STATE_STENCIL_WRITE_MASK:          return StencilWriteMask;
+		case VK_DYNAMIC_STATE_VERTEX_INPUT_BINDING_STRIDE: return VertexStride;
+		case VK_DYNAMIC_STATE_VIEWPORT:                    return Viewports;
+		case VK_DYNAMIC_STATE_VIEWPORT_WITH_COUNT:         return Viewports;
+		default:                                           return Unknown;
+	}
+}
+
+// This is executed first during pipeline creation. Do not depend on any internal state here.
+void MVKGraphicsPipeline::initDynamicState(const VkGraphicsPipelineCreateInfo* pCreateInfo) {
+	const auto* pDS = pCreateInfo->pDynamicState;
+	if ( !pDS ) { return; }
+
+	for (uint32_t i = 0; i < pDS->dynamicStateCount; i++) {
+		auto dynStateType = getRenderStateType(pDS->pDynamicStates[i]);
+		bool isDynamic = true;
+
+		// Some dynamic states have other restrictions
+		switch (dynStateType) {
+			case VertexStride:
+				isDynamic = _device->_pMetalFeatures->dynamicVertexStride;
+				if ( !isDynamic ) { setConfigurationResult(reportError(VK_ERROR_FEATURE_NOT_PRESENT, "This device and platform does not support VK_DYNAMIC_STATE_VERTEX_INPUT_BINDING_STRIDE (macOS 14.0 or iOS/tvOS 17.0, plus either Apple4 or Mac2 GPU).")); }
```

Comment:
This check on the other hand is probably necessary because there's no way for the validation layer to catch it. Maybe we need another feature in the `VkPhysicalDevicePortabilityFeaturesKHR` struct.

Reply:
> Maybe we need another feature in the VkPhysicalDevicePortabilityFeaturesKHR struct.

Yes. My thinking exactly. For the Vulkan 1.3 version of that struct.